rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,487 @@
+This architecture scales beyond two locations, with the metastore remaining the
+bottleneck. Therefore, finding a high-performance and scalable metastore is of
+essence.
+
+#### Example: Load-based scaling
+
+When VAST is deployed on top of an auto scaling engine, like Kubernetes or
+Nomad, we naturally model the components that are concerned with read and write
+paths into separate scaling groups. In this scenario new instances come and go
+depending on demand. The elasticity of the write side is automatically handled
+with the architecture, because there is no data sharing between writing
+instances, and the multiple ingest problem is handled in the same way as for the
+other scenarios.
+
+For the read side we need the ability to start additional nodes quickly, which
+adds the requirement for a quick resynchronization ability to the metastore.
+
+The most extreme variant of this scenario would be to create a dedicated
+instance for each query, which can be realized on top of serverless
+architectures such as AWS Lambda or Azure Cloud Functions. However, for systems
+with an at-least medium query rate we recommend to use longer-running read nodes
+to exploit IO synergies and in-memory caches.
+
+## Implementation
+
+We briefly touch on the implementation of write and read path, before dedicating
+the majority of attention to the metastore
+
+### Write Path
+
+To achieve fault-tolerance on the write path, we need an acknowledgement
+mechanism to confirm whether events at the source have been successfully
+processed. The current architecture looks as follows:
+
+![Write Path](write-path.png)
+
+We are assuming that we can begin with reading from a reliable medium that
+supports an indexed stream of events, such as Apache Kafka. If no such data
+acquisition mechanism is available, we may consider a local disk buffer to
+emulate a reliable medium. For this proposal, working around an unreliable data
+transport channel is out of scope.
+
+A transaction that writes one or more partitions needs to store additional
+metadata, containing the offsets this partition is based on. This atomic commit
+of both the transaction and its source position allows the source to pick up
+exactly at the point of failure, without risk of re-ingesting twice or missing
+any events. It thus provides exactly-once semantics for our write path.
+
+In technical terms, we can extend the CAF Streaming message type to include a
+handle to the source and source specific position, and then trigger an ACK once
+the data has been persisted. VAST already implements a similar mechanism today
+that acknowledges when events arrived at an active partition, which (almost)
+guarantees that they are queryable.
+
+When an exising source reconnects to an index component after a fault, it
+needs to request the most recently acknowlegded source position from the
+catalog and advance to the returned position to prevent double ingestion.
```

Comment:
Actually I see the skipping on the client side of events that were already ingested as an optimization. The really crucial mechanism is for the transactional component to discard the events that were already ingested.